src/clusters.rs

```diff
@@ -0,0 +1,940 @@
+                last_xform: 0,
+                last_color_xform: 0
+            };
+
+            points.push(point);
+        }
+
+        points
+    }
+
+    fn len(&self) -> usize {
+        self.num_points
+    }
+}
+
+/// A 2D quad of evenly-spaced points.
+pub struct GridQuad {
+    /// Center of the quad
+    center: Vec3,
+    /// x-axis of the quad. You're going to make this a unit vector, right? ;)
+    x_dir: Vec3,
+    /// y-axis of the quad. You're going to make this a unit vector orthogonal to x_dir, right? ;)
+    y_dir: Vec3,
+    /// Width of the quad in the x direction
+    width: f64,
+    /// height of the quad in the y-direction
+    height: f64,
+    /// The quad starts with a solid color
+    color: Vec3,
+    /// Generate up to N points (possibly less) in the grid
+    num_points: usize
+}
+
+/// Parse a GridQuad generator from JSON of the form:
+/// ```text
+/// {
+///     "type": "quad",
+///     "center": [x, y, z],
+///     "dims": [width, height],
+///     "x_dir": [xx, xy, xz],
+///     "y_dir": [yx, yy, yz],
+///     "color": [r, g, b] // 0.0 to 1.0
+///     "num_points": N
+/// }
+/// ```
+impl GridQuad {
+    pub fn from_json(json: &JsonValue) -> Self {
+        let center = Vec3::from_json(&json["center"], Vec3::zero());
+        let dims = &json["dims"];
+        let width = dims[0].as_f64().unwrap_or(1.0);
+        let height = dims[1].as_f64().unwrap_or(1.0);
+        let x_dir = Vec3::from_json(&json["x_dir"], Vec3::new(1.0, 0.0, 0.0));
+        let y_dir = Vec3::from_json(&json["y_dir"], Vec3::new(0.0, 1.0, 0.0));
+        let color = Vec3::from_json(&json["color"], Vec3::ones());
+        let num_points = &json["num_points"]
+            .as_usize()
+            .expect("num_points must be a positive integer");
+
+        Self {
+            center,
+            x_dir,
+            y_dir,
+            width,
+            height,
+            num_points: *num_points,
+            color
+        }
+    }
+
+    to_box!(Cluster);
+}
+
+impl Cluster for GridQuad {
+    fn generate(&mut self, set_id: u16) -> Vec<InternalPoint> {
+        let mut grid = Vec::new();
+
+        // Compute the effective grid size. If n points cannot be evenly
+        // divided into the same ratio as width/height, the grid may be
+        // slightly smaller
+        // See https://www.desmos.com/calculator/ltvzx6ezec
+        let n = self.num_points as f64;
+        let area = self.width * self.height;
+        let density = n / area;
+        let sqrt_density = density.sqrt();
+        let x_count = (self.width * sqrt_density).floor() as usize;
+        let y_count = (self.height * sqrt_density).floor() as usize;
+        let m = x_count * y_count;
+
+        let color = HalfMultivector::from_vec3(&self.color);
+
+        for i in 0..m {
+            let row = i / x_count;
+            let col = i % x_count;
+            let u = (col as f64) / ((x_count - 1) as f64);
+            let v = (row as f64) / ((y_count - 1) as f64);
```

Comment:
Hm, it would be really nice if I could store the UVs in the initial point for more shader magic.

Though this could be its own PR.